scripts/gen/tpcc.sh

```diff
@@ -118,33 +171,154 @@ then
   SET CLUSTER SETTING kv.snapshot_recovery.max_rate = '128 MiB';
   SET CLUSTER SETTING kv.snapshot_rebalance.max_rate = '128 MiB';
   ";
-
-  echo "importing..."
+  echo "importing ..."
   ./cockroach workload fixtures import tpcc --warehouses="$f_warehouses" --active-warehouses="$f_active" "${pgurls[0]}"
   echo "done importing"
 fi
 
-# if f_inc is not specified, then we set it equal to f_warehouses, this means
-# we just need to run one iteration.
-if [[ $f_inc == 0 ]];
+# Comment out since we don't have a scenario using this at present.
+## if f_inc is not specified, then we set it equal to f_warehouses, this means
+## we just need to run one iteration.
+#if [[ $f_inc == 0 ]];
+#then
+#  f_inc=$f_warehouses
+#fi
+
+# We first run an exploration test round with binary search to find a right
+# range of test boundary to use in the subsequent official run.
+# For certain machine types the binary search result don't work well and will
+# cause kv errors and end the test, so we need to handle them differently.
+tpcc_files=""
+has_good_result=0
+if [[ special_type -eq 0 ]];
 then
-  f_inc=$f_warehouses
+  l=$f_active
+  r=$f_warehouses
+  while [[ $l -le $r ]]
```

Comment:
When `l == r`, we end up running `f_warehouses + f_inc` which will likely fail since `f_warehouses` is intended to be an upper bound, right?

Reply:
`l==r` is just the right formula to end the search, the real boundary of the subsequent run is controlled by the `best_result`, then boundary is set around this sweet spot after adjustment.